### ISSUE TITLE
Need more intelligent re-resolution of names

### ISSUE THREAD

**Issue by carldanley** [grpc/grpc#12295]
For starters, let me state that this is not a bug. I'm trying to reach out for help because I can't simply piece together enough information to figure out how to address client-side load balancing properly within the node.js world of gRPC.

Here is my example code base: https://gist.github.com/carldanley/39d5a0d7f9b1ea865af94481da1e0cac. I deploy that to a kubernetes environment and use a load balancer to *attempt* to split the traffic (with no luck)... What am I doing wrong?

**Comment by markdroth** [comment]
For future reference, for items like this that are not bugs, instead of filing issues, please send email to the grpc.io mailing list.  That's a much better forum for this kind of discussion.

I don't know anything about node or speak javascript, so @murgatroid99 will have to help you with that side of things.  But I can help answer your questions about how client-side load balancing is supposed to work.

I think a similar question came up a while back; see discussion in #11406.  Glancing at your code, it looks like you're using the wrong name for the channel arg to select the LB policy.  Try changing `loadBalancingPolicy` to `grpc.lb_policy_name`.

If that doesn't fix the problem, please let us know what you're trying to do, what you expected to see, and what you're actually seeing.

**Comment by carldanley** [comment]
Hello @markdroth! First off, thank you!

Secondly, I tried changing the load balancing policy to `grpc.lb_policy_name` but had no luck with client-side load balancing occurring. Let me explain what I'm using for a setup:

I run kubernetes. I have a service for our gRPC server. I have a client that is told to use the internal DNS address for the service: `server.grpc.svc.cluster.local` which resolves to the 2 (for testing purposes) pods that run the gRPC server. When the client is instructed to connect to that service (via https://gist.github.com/carldanley/39d5a0d7f9b1ea865af94481da1e0cac#file-index-js-L50), it connects to the first one that the DNS resolves to and keeps creating gRPC requests to that single server (instead of round robin to the two servers). I'm not sure what I can do differently to make this work... I expect the code in that gist to have a single client connect to both hosts on start and then, per request, round-robin through each of the gRPC servers.

**Comment by carldanley** [comment]
Ok, so this was a kubernetes-related issue. I had to turn off kubernetes load balancing for this specific server's service entry in k8s. However, the next thing I tested was destroying a server randomly. It appears that the client does not periodically re-evaluate the entries for the DNS name provided; as such it continued to send traffic to the remaining instance it knew of but never detected the new instance (until I restarted the client). How can I have my clients automatically start using a new server instance that came online?

**Comment by markdroth** [comment]
Currently, we only re-resolve DNS names when all subchannels become disconnected.  So if you restart both of the servers, the client will re-resolve and then connect to the new addresses.  But if you move just one of the servers, we will just stop being able to use that one.

@dgquintas and I have talked about possible ways to address this problem.  It's fairly tricky, because we can't really know when it's useful to re-resolve, and we need to avoid slamming the DNS server with constant queries.  For example, if a server is crashing every 10 seconds, we don't want every single client to try to re-resolve the name every 10 seconds.  And if this is an environment where servers have static addresses, then there's no point in re-resolving in the first place.

One possible solution would be to make the DNS resolver aware of DNS TTLs, so that we can automatically re-resolve after the previous results expire; this would essentially allow the DNS data to determine how often the clients re-resolve.  However, while we could probably do this in the C-core gRPC implementation, it's not clear that we have reasonable ways to access DNS TTL information in Java or Go, which would make our clients' behavior inconsistent.

Another possibility is to provide the ability to configure the threshold for what percentage of subchannels need to become disconnected before we re-resolve.  The default would be 100%, which would match the current behavior, but it would allow people to reduce the threshold to something more appropriate for their environment.  We might also want to provide a way to set the minimum interval between re-resolutions, just to provide some additional safety against slamming the DNS server.

Anyway, we've had a lot of discussions about this but have not yet decided on any particular behavior or scheduled any work on this.  But if this is something you'd like, let us know, and we can start figuring out how to prioritize it.

**Comment by carldanley** [comment]
Okay, I understand. I did some testing and splitting traffic reliably in a CI/CD environment is hit or miss at the moment. Consider the following:

You have 2 instances (iA and iB) of service 1 (s1) running. You have 2 instances (iC and iD) of service 2 (s2) running. If s1 has a rolling update and shortly after s2 has a rolling update: the events could pan out like this:

1. `s1 rolling update starts`
1. `iA goes down`
1. `iA comes up`
1. `s2 rolling update starts`
1. `iB goes down`
1. `iC goes down`
1. `iB comes up`
1. `iC comes up`
1. `s1 rolling update stops`
1. `iD goes down`
1. `iD comes up`
1. `s2 rolling update stops`

In the scenario above, s1 instances (iA and iB) will only see 1/2 of s2 instances (iC) given. This is bad and means that none of my traffic is round-robin'ed and that 1 instance is getting slammed.

After reading the solutions you proposed, it kind of feels like all of those things should at least be available. Let the developer decided whether or not they want to take extra DNS traffic. Let the developer decided if they're using a client (in a language) which can access DNS TTL and so on.

*Just thinking out loud here:*

I think there are really 2 scenarios we care about:

**1. Losing an established subchannel**

When a subchannel disconnects (2/2 instances becomes 1/2 instances), we could attempt to re-resolve the DNS entry for some number (attempt-based or time-based) of times. This gives us a way to start listening until we're 2/2 again OR we were unsuccessful in getting back 2/2 (so we stay 1/2).

**2. Discovery of new instances**

Consider we had a healthy, happy service that was correctly load balancing 5/5 instances. What if we had autoscaling enabled and we had a sudden surge of traffic hit our servers. Now we're running 7 instances and because we never scaled the clients and only the servers, we have no way to serve the surge of traffic (because they'll never refresh the DNS pool) so we stay 5/7...

One possible solution is giving us a function to call that can refresh the pool of hosts via DNS resolution. This would give us a way to decide when we want to trigger it ourselves. For example, imagine that we performed a rolling update on a service that had 100 instances; we could publish an event onto our messaging queue (when 100 of 100 is up) that could tell clients to refresh their DNS hosts. Really, we could write whatever logic we want with this and perform the resolution whenever we saw fit; it gives us full control of something that works for us.

Anyways... </2cents>

**Comment by markdroth** [comment]
I definitely agree that we should do better in the first scenario you mentioned, and some combination of the ideas we've been discussing could address that.

With regard to the second scenario you mention, I think it's worth noting that DNS is fundamentally unsuited to the kind of dynamic environment you're describing, because DNS is a polling-based mechanism, whereas what you really want is a push-based mechanism where the clients are proactively notified when addresses change.  While we might be able to find a way to work around this with DNS with the DNS TTL solution I mentioned above, I think it will never really scale the way it needs to, because it really wasn't designed for this kind of usage.  A better approach would be to write a new resolver mechanism that subscribes to notification from some centralized system as the servers move around.  For example, I'm not sure what mechanism kubernetes uses to update DNS, but you could presumably have it also notify some other name service that would allow clients to subscribe to particular names and would proactively send them updates when kubernetes notifies them of changes to those names.  Then your clients would be getting a constant stream of updates and would always have an up-to-date set of addresses.

Given that, I think that any changes we make here will likely be focused on the first scenario, not the second.  But we'll have to talk further to decide exactly how we're going to handle this.

**Comment by hollinwilkins** [comment]
@markdroth DNS may not be the perfect solution, but it is ubiquitous and easy to integrate with. I would prefer to setup a DNS poll every 10-20 seconds for my microservices to at least get going with load balancing my gRPC services. When that produces too much load on the DNS servers, then I will start looking at a lookaside balancer.

Right now the cost to getting simple load balancing that we are used to with HTTP 1.1 is very high. The solutions are, as I see them:

1. Create your own lookaside load balancer, modify ALL of your clients code to use the lookaside balancer, maintain the code and integrations with your service discovery platform of choice (K8, Consul, ZooKeeper, etc.)
2. Create a service mesh using Istio or Linkerd, both of which have their own limitations, drawbacks, and advantages.
3. Use builtin DNS resolver, which means you can't scale your servers up without first scaling all of your servers down.

A DNS-based resolver with a refresh interval would be a very low-cost, low barrier-to-entry solution that lots of developers would be comfortable with and not require a huge investment in either infrastructure or coding.

**Comment by hollinwilkins** [comment]
For anyone encountering issues and looking for a simple solution: https://github.com/grpc/proposal/pull/23/files

Using server-side connection options can cause load to redistribute in a fairly easy manner! Wish I had seen this document 2 weeks ago.

**Comment by jtattermusch** [comment]
CC @jtattermusch

**Comment by jtattermusch** [comment]
@hollinwilkins  can you describe what changes to your setup you've made (in reference to https://github.com/grpc/grpc/issues/12295#issuecomment-358483266) and confirm that the "discovering of new endpoints" problem went away?   I am currently facing the exact issue you were facing (losing an established instance is handled correctly, but new instances are not being discovered) while trying to make a simple RoundRobin LB scenario work out of the box on kubernetes.

Are there any other possible workarounds (like forcing re-resolution of backends)?

**Comment by jtattermusch** [comment]
Ad workaround based on https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md:

I tried setting `grpc.max_connection_age_ms` and `grpc.max_connection_age_grace_ms` channel arguments on the server I'm trying to access with RoundRobin load balancing policy and it seems that it is helping
- closing the connections occasionally leads to re-resolving the domain name and newly added service replicas are being picked up by the round-robin loadbalancer in a relatively short time.

**Comment by hollinwilkins** [comment]
@jtattermusch This is the approach I took. Not ideal, but works for now.

**Comment by wjywbs** [comment]
grpc go client re-resolves DNS every 30 minutes. Could c++ client do the same, so we can configure the interval?

https://github.com/grpc/grpc-go/blob/master/resolver/dns/dns_resolver.go#L46

**Comment by hacst** [comment]
After having read this thread and some of the linked issues I'm still not sure I understand why observing the DNS TTL for refresh would be a bad thing. From what I can tell it would just work. Be it scaling up or down, k8s or outside of it. I think its properties would fit the principle of least surprise. I cannot imagine many selecting a DNS based round-robin load balancing approach would be surprised by clients having to poll DNS in TTL interval and that producing load. However many will be surprised to learn it won't react to changes in DNS.

Load seems to be the most commonly stated reason why observing DNS TTL would be bad but I just don't see it. If my DNS service cannot handle the polling load I can easily trade-off with higher TTL, scale my DNS and ultimately once that no longer makes sense transition to another more scaleable LB approach. It is not like DNS RR LB in gRPC allows arbitrary scale to begin with so why pretend it has to? It is the simple solution for the simple cases. It should work as best as it can inside of those constraints.

Having to use MaxConnectionAge, which just happens to be coupled to re-resolution, to emulate a polling behaviour seems like a bad workaround to me. I don't see how making a DNS query to some DNS cache every X seconds would be seen as problematic but having to do a magnitudes more expensive re-connect plus (encryption-)handshake with each of the backends plus having to regularly refresh DNS anyway is an acceptable workaround to that.

Currently all the alternatives I can see are vastly more complex to run and expensive to implement. Why force users to use a service mesh or some custom look-aside load-balancing scheme when there's a way to make what is already supported just work for a lot of cases?

**Comment by markdroth** [comment]
Personally, I tend to agree that the max-connection-age approach is a fairly ugly solution to the problem of forcing re-resolution.  However, I'm not sure that everyone on the team agrees with that.

I think the main argument against using TTLs is that we want consistent client behavior across languages, but while we would be able to access the TTL information in C-core, we have no reasonable mechanism for doing so in Java or Go.  So it's not really a portable solution.

I do think we should consider providing a way for the client to be configured to periodically re-resolve at some fixed interval.

I'd like to get feedback from @ejona86, @zhangkun83, and @dfawley on this.

**Comment by hacst** [comment]
I see. Technically DNS TTL could be retrieved in any language by using a custom resolver (e.g. using something like miekg/dns in go or netty DNSResolver in Java. The latter would also get rid of the broken built-in DNS caching behaviour of the JVM...). That's basically what using c-ares in C-core amounts to. Whether that's a "reasonable" thing to do everywhere is of course debatable.

In any case I definitely would prefer a configurable polling interval for DNS to the current MaxConnectionAge approach. Maybe there could even be an opt-in flag that makes it use the DNS TTL when supported and fallback to the polling interval otherwise? I'm not sure whether such "extensions" is something that's done across the gRPC clients in different languages but I would be surprised if they are totally equal now. But as I said. Just having the configurable DNS polling interval would be a considerable improvement.

**Comment by hacst** [comment]
@ejona86 I see your point about the stampeding herd. I hadn't considered that each cache will return the current remaining TTL of its cached value (which is kinda obvious in hindsight) so any client talking to that cache instead of the authoritative source will sync up when polling. That definitely isn't a great behaviour if you want to have thousands of clients.

While it isn't as bad having these thousands of clients reconnecting to each single backend server with MaxConnectionAge period doesn't sound great to me either.

Besides pure re-connection cost, tearing down perfectly fine connections that could otherwise stay long-running can also have other side-effects at a higher level in the stack. E.g. assume a service offers very long-running bidir streaming calls with an expensive to re-create context on the server related to the running call. In that case using MaxConnectionAge will force the client to regularly end the call and disconnect. The next call will hit some other random backend in which the context has to be re-created.

Could you elaborate on the configuration issues you see with a configurable DNS polling interval disconnected from TTL? I would've thought it would just be a value like MaxConnectionAge that does nothing if not set.

**Comment by ejona86** [comment]
> assume a service offers very long-running bidir streaming calls with an expensive to re-create context on the server related to the running call. In that case using MaxConnectionAge will force the client to regularly end the call and disconnect. The next call will hit some other random backend in which the context has to be re-created.

So two parts to this:
1. MaxConnectionAge itself doesn't require the stream to be torn down. MaxConnectionAgeGrace will control how long old streams can live. Old streams can stay on old connection indefinitely, but will also keep the old connection alive as long as they do so. In the worst-case, over time each old stream will basically each get its own connection if streams have infinite lifetime. But we tend to expect few long-lived streams per backend.
2. We've seen that service-owners frequently need to put a lifetime on streams, otherwise they aren't load balanced. If you bring up a new server, none of the existing clients will create the long-lived stream to the new server because they are happily connected to an old server.

It is possible to develop a client-side LB policy that uses affinity to consistently route long-lived streams back to the same warm backends, but because of (2) it puts you in a bit of a bind for distributing load when applied to this use-case. (This affinity-based system actually exists in gRPC Java today but the design went into a weird limbo state as we resolve some larger LB discussions. It was implemented in Go as well, but was reverted because the design went into limbo state. It is powered off service config, which isn't ready for prime-time, though.)

> Could you elaborate on the configuration issues you see with a configurable DNS polling interval disconnected from TTL? I would've thought it would just be a value like MaxConnectionAge that does nothing if not set.

MaxConnectionAge is configured on server-side. So if the service owner needs to change the value, they can change it fairly rapidly. Most obvious forms of configuring DNS polling interval would place it hard-coded on client-side, which means it can take O(years) for clients to pick up any change. Yes, some service owners control their clients and so it wouldn't be a problem, but many don't and so the solution would have more limited applicability.

While it could be possible to provide live configuration to the client via the service config, that necessitates increasing the complexity of the solution.

**Comment by srini100** [comment]
In theory, the timing of connection aging can line up in such a way that all backends drop connections almost at the same time causing increased latency spikes on clients even in a steady-state system. This would not be the case with periodic DNS refresh. 

Both solutions (periodic DNS refresh and max connection age) have pros and cons. Are there are any users who have run into practical issues with max connection age solution?

**Comment by schmohlio** [comment]
bumping this - couldn't MaxConnectionAge approach also lead to a "stampeding herd"?  Also curious if anyone has experienced issues with the connection-based approach.

We would like to make use of out-of-the-box autoscaling features, whereby DNS records are added when new boxes come online. It's preferable that we don't need to bounce a service or set MaxConnectionAge (as this reduces the responsiveness of overall autoscaling approach).

**Comment by ejona86** [comment]
> couldn't MaxConnectionAge approach also lead to a "stampeding herd"?

No, because it is based on the original connection creation. It can perpetuate a stampeding herd (by making it reoccur), but wouldn't be the cause. This is also why it uses a 10% randomization factor.

> It's preferable that we don't need to bounce a service or set MaxConnectionAge (as this reduces the responsiveness of overall autoscaling approach).

Services using DNS today create new connections _very_ frequently. If you want a MaxConnectionAge of 30 seconds, that's okay (although low ages may be more problematic for Go clients). The connection has been utilized _much_ better than it would have with HTTP/1 and the cost has likely been amortized over many RPCs. HTTP servers frequently shut down connections much younger.

(Go clients currently reconnect eagerly. So with a MaxConnectionAge of 30 seconds the clients will reconnect every 30 seconds, even if they only do an RPC once an hour. This is a TODO for Go, and doesn't impact many users.)

**Comment by wjywbs** [comment]
I tried to enable MaxConnectionAge on a server and call from a Go client. When the connection age was reached, the rpcs in the Go client failed with the unavailable error and the "transport is closing" message. Then I gave up on this MaxConnectionAge approach to avoid extra retry logics in the code.

**Comment by ejona86** [comment]
@wjywbs, is there a bug open for that? That should not happen. @dfawley, do you know what could tickle the Go client to do that?

**Comment by dfawley** [comment]
The only thing I can think is if you have long-running RPCs, max connection age will eventually time out (after `MaxConnectionAgeGrace`) and hard-close the connection.  Otherwise the RPCs are expected to complete successfully, and I'd be pretty surprised if we don't have tests for this.  If the above doesn't explain the errors you're getting, please file a bug in the grpc-go repo about that.

**Comment by wjywbs** [comment]
Thanks for your help. I didn't set the grace time last time. However, when I tested again with both age and grace time set to one minute, the Go client reported lots of errors as well. Each rpc takes a few seconds to complete, within the one minute grace period.

```
grpc-go v1.18.0/v1.19.1
rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: <nil>
rpc error: code = Unavailable desc = the connection is draining
rpc error: code = Unavailable desc = transport is closing
```

**Comment by ejona86** [comment]
@wjywbs, please file that as an issue on grpc-go's repo. That does not appear to be expected behavior.

**Comment by rantav** [comment]
Hi, I wanted to chime in with a different suggestion to resolve similar problems.   
My suggestion is discussed here https://github.com/grpc/grpc/issues/18743 (closed by @markdroth as dup of this issue) and here https://github.com/grpc/grpc-go/issues/2751 but let me reiterate the highlights for brevity. 

I suggest to configure the DNS resolver using query parameters in the connection URL. This is similar to how JDBC, for example, works.  
Example: 

```
dns://127.0.0.1:8600/endpoint.service.consul.:8080?lookupSRV=0&lookupTXT=0&refreshRate=30
```

Where:
* `127.0.0.1` is the IP of the authority DNS server (this is not new)
* `8600` is the DNS server port (default is 53, this is not new)
* `endpoint.service.consul.:8080` is the endpoint:port to lookup (this is not new)
* `lookupSRV=0` disables SRV records lookups (this is new)
* `lookupTXT=0` disables TXT records lookup (this is new)
* `refreshRate=30` configures the refresh frequency to 30 seconds (this is new)

My suggestion is to be able to declaratively configure the DNS **refresh rate** (also called `minFreq` as in minimal lookup frequency), whether to **lookup SRV** records or not (default to true but not always necessary and when lookups are often, this can load on the DNS server) and whether to **lookup TXT** records or not (likewise default to true but not always necessary and can load)

**Comment by ejona86** [comment]
We have discussed query parameters before. I'm not against them in general. The main issue with query parameters _for DNS refresh rate_ is that the service owner has no control of the setting. The main issue of doing DNS refresh rate is that we don't want a client-side option, and making it service-controlled has complexity.

Let's not bring SRV and TXT into this discussion. Although I will note there is a channel option to disable service config lookup (TXT), independent of the name resolver. In C it is GRPC_ARG_SERVICE_CONFIG_DISABLE_RESOLUTION, but it will exist in every language (Java got it in v1.20).

**Comment by rantav** [comment]
@ejona86 thanks for your feedback.  The reason I brought SRV and TXT to this discussion is that when refresh rate is high and either SRV and TXT are enabled then the load on DNS could potentially become much higher. I do see your point isolating the topics however I wanted to mention the reason for bringing them in.  

**What then would be the best way to proceed from here?**

Let me just restate the use case I'm interested in (I believe it's a common scenario, I just want to be explicit). 

### Use case:
Using the grpc DNS resolver; obtaining several IPs from the DNS server. 
1. Then one of the serves goes down (malfunction or deployment). The sub-channel shuts down cleanly (from what I observe) and this is good
2. However when new servers come up (new deployment or scale up), then DNS resolver would not notice them until re-resolution occurs, which is only after 30 minutes (in Go). 

2 is the problem. We want the grpc clients to notice the new servers as soon as possible. We can tolerate maybe 30 seconds delay but not 30 minutes.

**Comment by ejona86** [comment]
The closing of a connection triggers a DNS re-resolution. Although I will say that the DNS re-resolution generally occurs in parallel with re-connecting, so the DNS resolution may complete _after_ a new connection is created.

A rolling restart would work fine. However, if you are scaling up your backends then clients wouldn't actively disconnect from their current backend, so you would discover little load is redistributed. For that, specify MaxConnectionAge on the server-side, which will force the client to periodically close the connection and re-resolve.

**Comment by rantav** [comment]
@ejona86 thank you for the suggestion. I suspect, though (based on my limited testing and code reading) that it is suboptimal. 

>  if you are scaling up your backends then clients wouldn't actively disconnect from their current backend, so you would discover little load is redistributed. For that, specify MaxConnectionAge on the server-side, which will force the client to periodically close the connection and re-resolve.

Using `MaxConnectionAge` and setting it to, say `30s`, would work as you explain, clients would disconnect and then dns resolution would occur and then new backends would get discovered.  
The downside of this workaround (sorry, I see it as a workaround) is that **clients would disconnect**. They would disconnect from healthy backends every 30 seconds regardless of new backends scaled up or what.  
At high scale when multiple clients are connected to multiple backends the performance hit could become non-negligible.  

In my suggested solution which includes DNS re-resolution (based on `refreshRate`) this would not happen. Existing sub-channels would not get disconnected. 

I could use this as a workaround for now, but I think it's sub-optimal. 

I tested in grpc-go

**Comment by ejona86** [comment]
> The downside of this workaround (sorry, I see it as a workaround) is that **clients would disconnect**.

Yes. And in the pre-existing systems using DNS like this, _every request_ frequently creates a separate connection. And even if a connection is re-used, it has a pretty low lifetime (definitely < 1 minute).

> then any client would in turn run dns re-resolution every 3s. At high scale when multiple clients are connected to multiple backends the performance hit and the load on DNS could become non-negligible.

Pre-existing clients are _already_ doing stuff like this, each time they create a connection. I'm only aware of custom client-side LB policies that would do anything differently.

But gRPC does delay/combine DNS requests if one has happened in the last X seconds. See https://github.com/grpc/grpc/pull/14170#issuecomment-360563827 . That has the effect that new clients will have two polls at roughly MAX_CONNECTION_AGE frequency (the two polls will be X seconds apart), because all the connections will age out around the same time. Over time the reconnections will drift and so the client will re-resolve at a rate of `MAX(X seconds, MAX_CONNECTION_AGE/backends)`. You are free to account for this on the server-side and choose a higher MAX_CONNECTION_AGE; yes not _all_ clients would notice new backends as quickly, but a large enough majority could notice to spread the majority of the load quickly.

The gRPC solution isn't "optimal," but it is better than what most systems I've seen have been doing. The solution here is simple, leverages features necessary for other cases like L4 LBs, and keeps the service owner in control of the setting. If you want clients to notice new servers very soon, use a name resolution system that has notifications. That's the only "optimal" solution.

**Comment by markdroth** [comment]
I don't think we ever reached a resolution on this.  This should stay open until we decide once and for all how we actually want to deal with this.

**Comment by gmehrotr** [comment]
Are there any plans to provide a fix for the current C++ gRPC library behavior to not honor DNS TTL and rather refresh the DNS cache only when the initial attempt to connect to a stale IP has failed ?

I tried following the discussions above but its too confusing and unclear whether there is a solution or not.

**Comment by markdroth** [comment]
The current state is that gRPC does not honor TTLs nor does it periodically refresh.  It will only refresh when a backend connection fails.  So the current recommendation is to use the server-side connection management to force clients to reconnect periodically, which will also force them to re-resolve.

It's not a great solution, and a lot of people would like the ability to periodically refresh, but we have not managed to reach consensus on making that happen.

**Comment by gmehrotr** [comment]
Thanks!
Can you help elaborate further on server side connection management to force client to reconnect ?

We are using AWS load balancer  with connection idle timeout of 60 sec but grpc client continues to use stale IP address of load balancer even without attempting DNS resolution which causes intermittent failures each time the load balancer IP address is changed.

Our goal is for the client to try DNS resolution for each new request such that it can use one of the fresh IPs for LB for making the connection and avoid the failure.

**Comment by markdroth** [comment]
The server-side connection management allows you to have your gRPC server disconnect after some maximum period of time has been reached.  For details, see:

https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md

We do not have a way to have the client try DNS resolution for each new request.  That seems like it would cause you to DoS your DNS server.

**Comment by gmehrotr** [comment]
Sorry, I don't follow what you mean by gRPC server ?

In our case we are using c++ gRPC client to connect to Google Cloud TTS API via proxy server.

gRPC Client --> AWS LB-->Proxy server-->Google Cloud TTS Endpoint

As I said before, even though AWS LB has max idle timeout set to 60 seconds the gRPC client keeps using stale IP to connect to it for subsequent requests.

**Comment by markdroth** [comment]
I don't think I understand enough about your setup to know how to advise you.  If the AWS LB is dropping the connection to the client, that should cause the client to re-resolve.

**Comment by gmehrotr** [comment]
@markdroth AWS LB currently does drop the connection from client after 60 sec but still client does not re-resolve.

Currently our implementation creates a gRPC stub only once during startup and continues to re-use the stub for handling subsequent Google TTS requests which keep using stale IP address for LB (without any DNS resolution).

So is the DNS resolution done only when we create a new gRPC stub ?

**Comment by markdroth** [comment]
No, the DNS resolution is done every time the client sees a disconnect.  It has nothing to do with stub creation.

Note that when the backend disconnects, the client will kick off a re-resolution, but it will not wait for the re-resolution to complete before it attempts to reconnect to the original addresses.  If the server has actually moved to a new address, then the attempt to reconnect to the original address will fail, and typically the re-resolution will have completed by the time that connection attempt fails, at which point the client will attempt to connect to the new address.  However, if the server at the original address is still answering incoming connections, the client will wind up using that connection, even if re-resolution completes a moment after the connection is established.

It's a fairly unusual situation to have DNS change the address but have the old address still answering and not being able to function.  Typically, a server will not answer if it's not going to function.  But if you think that's what's happening here, feel free to file a separate issue.

**Comment by ejona86** [comment]
> However, if the server at the original address is still answering incoming connections, the client will wind up using that connection, even if re-resolution completes a moment after the connection is established.

Although in this case, when DNS does respond we will reconnect (again) if the address we are using is no longer in the list of addresses.

**Comment by markdroth** [comment]
> Although in this case, when DNS does respond we will reconnect (again) if the address we are using is no longer in the list of addresses.

Yes, but any call that was started on the old address before the re-resolution completed will have failed.

Also, I think in C-core, we currently don't switch over to the new addresses until one of them reports READY.  That is something we should fix, though.

**Comment by gmehrotr** [comment]
Thanks Mark. We will investigate further why we don't see resolution happening each time even though Load balancer closes connection after 60 sec idle timeout.

For context - We are using AWS classic load balancer and quite frequently we see the ip address of the backend pool change while the client keeps using the old stale IP(not even in the DNS) which causes the request to fail. Once this happens, DNS cache is refreshed and subsequent request pass until we encounter the IP address situation once again.

**Comment by ejona86** [comment]
> Yes, but any call that was started on the old address before the re-resolution completed will have failed.

They shouldn't. In Java they don't. It should be treated similar to a GOAWAY where the old calls continue on old connection and new calls use the new address/connection.

**Comment by markdroth** [comment]
Eric, that is the behavior.  But in this situation, it sounds like the server at the old address is still there and still responding, but the requests will no longer succeed.  From gRPC's perspective, that's your "old calls continue on old connections" case, but it just so happens that in this situation, those calls will fail.

**Comment by dmarkhas** [comment]
> It's not quite that simple. When caching DNS resolvers are in place a single response from the authoritative DNS server can be sent to 1000s of clients. All those clients will have the TTL expire at the same time (independent of when they originally queried) so they form a "stampeding herd." Every time the TTL expires the entire "herd" will re-request DNS at the same time. Increasing the TTL would decrease average load but wouldn't reduce peak load.

Couldn't this be solved by introducing sufficient "jitter" that would use the TTL as a baseline and would refresh randomly between (for example) `[MinRefreshInterval, TTL]`?
MinRefreshInterval could be derived from the TTL itself, or configurable by the client.

**Comment by ejona86** [comment]
@dmarkhas, a jitter is much better than without, but it is a bit "chaotic." Although as I said a while back, I would like to focus on what _problems_ are caused by the current suggestion of using MAX_CONNECTION_AGE.

The client doesn't see the TTL configured on the DNS record. Instead, it sees the "current" TTL for the _cached_ record. So if the record started with 30s TTL, a client requesting the record from a caching DNS resolver 20s after the record was initially retrieved will see a TTL of 10s. Now, if a client sees a 1s TTL it will refresh very soon, but note it will get ~30s TTL for the second request.

Assuming the jitter is a uniform distribution, there would still be a skewing of refreshes toward when the TTL expires. But it may be "good enough." My concern for load distribution at that point is that it is a bit too complicated to intuitively understand how it would behave in practice, so we'd need to get a better understanding (simulations/statistics) to avoid surprises. 

Note that polling also has other issues, like some large deployments do not provide consistent results via DNS. Repeatedly run `dig google.com @8.8.8.8` as an example. This has worked without issue today because existing protocols like HTTP do not bind the lifetime of the connection to DNS results; instead, connections are re-created when they are needed and just use whatever DNS returns at that point. (I'm not saying that is ideal; many existing deployments have had serious trouble with how long it takes to drain traffic from an IP or toward a new IP.) So for these environments we don't want to swap to a new IP based on _polling_ DNS. We'd need to resolve how we'd handle these environments (e.g., opt-in to polling, "stop doing that" (consistent DNS results), etc.)

**Comment by dmarkhas** [comment]
I think the problem of using `MAX_CONNECTION_AGE` is that, while it achieves the end goal of allowing clients to discover new endpoints, it does so as a side-effect and at a very high, and unnecessary cost.
If the client wants to discover new endpoints periodically from DNS, it should be able to do so.
If the server wants to limit the clients' connection age, it should be able to do so.
But there is no reason that these two distinct operations would in any way depend on one another.

I agree that jitter is not as intuitive and transparent as a fixed polling period, but the safety it provides is worth this slight obscurity. It is used in many common applications that face similar problems, for example Hashicorp's Vault uses a combination of lease time (equivalent to DNS TTL) and a jitter to periodically refresh leases, without overwhelming the backend:

https://github.com/hashicorp/vault/blob/master/api/lifetime_watcher.go#L233

Your concern regarding load distribution is legitimate, as you are trying to stop people from shooting themselves in the foot with excessive DNS polling - but I don't think that simply taking away the option to poll DNS periodically is the right approach.
It can be disabled by default, but if someone understands the implications and potential impact, he should be able to turn it on. I think you should provide sensible defaults (i.e. no polling), and a fair warning for those who want to override those defaults, but you should not prevent them from enabling polling altogether.

**Comment by ejona86** [comment]
> I think the problem of using MAX_CONNECTION_AGE is that, while it achieves the end goal of allowing clients to discover new endpoints, it does so as a side-effect and at a very high, and unnecessary cost.

Is this based on experience, or gut feeling? I would love to see some evidence. Or even just an explanation of why your use case is particularly harmed by this approach.

As I said earlier, doing polling is an optimization. We need to make decisions about how much would be gained for what cost. I definitely believe some cases wouldn't be well served by the current approach. But I also believe that the current approach isn't as bad as many people feel it to be, especially given the HTTP/1 systems that many people are coming from.

**Comment by dmarkhas** [comment]
I don't have any metrics in regards of the overhead of the server dropping healthy connections. The argument against the current solution is probably leaning more towards unnecessary coupling between client code and server code, which in many cases cannot be easily reconciled since client and server can be developed by different teams, even different companies.

**Comment by ejona86** [comment]
@dmarkhas, the decoupling of the client vs server code is _exactly_ the reason why I recommend the current approach, as it allows the service owner to control scaling policies. You're saying in your case the _client_ is in a better position to know the necessary polling frequency instead of the service? That seems strange to me, especially since the service owner controls the DNS TTL. But if so, I'd really like to hear more about that.

**Comment by dmarkhas** [comment]
The client is not necessarily in a better position, but if the service owner did not implement this approach at all, it makes it impossible for the client to ever be able to discover new instances.

**Comment by ejona86** [comment]
@dmarkhas, are you saying you are in that position, or is it hypothetical? Server imbalance sounds like something the service owner would care about. If they don't, then either the client shouldn't care or the client owner should make them care by complaining about low QPS/high latency. If there are reasons the service owner can't easily enable something similar to MAX_CONNECTION_AGE, we could look into solving that. But if the service owner seriously doesn't care about their service's performance, spending effort on client-side seems like a lost cause.

**Comment by dmarkhas** [comment]
It's not entirely hypothetical, we've been in a similar situation recently, and it took a while for us to:
a) figure out why we're not load balancing our client connections across all instances that we see in DNS (due to this: https://github.com/grpc/grpc-go/pull/3228)
b) figure out how to get load balancing back, only to realize that we can't and we have to:
c) get the server-side team (located in a different site) engaged so they can turn on this setting on their end

As you can see, this is not as trivial as simply polling the DNS...

But I feel like we're kinda beating around the bush here. I'm not saying the existing solution is not viable, just that it is not as straight-forward as simply letting the client poll the DNS at his own leisure (which raises other concerns, but I feel that they can be sufficiently mitigated) 🤷

**Comment by ejona86** [comment]
If you didn't do (c), then all the other client teams would have to deal with (a) as well (Go was the only language with that behavior, so other languages would have already had trouble). Because you did (c), it is no longer a problem for other users. Having worked on clients quite a bit myself, I completely understand the pain in debugging server-side issues. But that doesn't mean any issue experienced on client-side should receive a client-side fix.

> I'm not saying the existing solution is not viable, just that it is not as straight-forward as simply letting the client poll the DNS at his own leisure

That's very fair. From my perspective this is not the client's job. I skew work from the client owner to the service owner, when possible. There are multiple orders of magnitude more clients, so it is best to avoid them needing any custom configuration, especially when it is necessary to maintain the health of the service. But it's also that if the clients hard-code configuration, it becomes a problem for the service owner as they may have to wait years/forever for their clients to upgrade to newer configuration.

@dmarkhas, it may be fair to add polling as an "escape hatch." The problem is I wouldn't have wanted _you_ to use that escape hatch! Except as a short-term workaround as you wait for server-side team. But if it is only a short-term workaround that helps in very specific situations (e.g., only one client team), it's unclear how much value it adds while opening the potential for misuse. And once you have a workaround in place, it is very common to stop pushing for a long-term solution.

**Comment by dmarkhas** [comment]
> From my perspective this is not the client's job. I skew work from the client owner to the service owner, when possible.

I admit that this was counter-intuitive for me as I was reading this discussion (and others) but you have helped shift my perspective and I understand your reasoning much better now.

> But it's also that if the clients hard-code configuration, it becomes a problem for the service owner as they may have to wait years/forever for their clients to upgrade to newer configuration.

That's true, but the flip side of it is clients waiting for the server-side team to upgrade to newer configuration :)

I think that if we allow clients to resolve DNS without being dependent on the server team, and the server team to drop connections (and force the client to resolve DNS) without being dependent on the client team, we will enjoy both worlds.

**Comment by aarontami** [comment]
We have been using the current solution of using MAX_CONNECTION_AGE on the server side and have run into a number of problems while doing this.  We started experiencing a large number of random UNAVAILABLE errors, which seemed to be due to the connection age and expiring and getting canceled while a request was in progress. I think we were able to solve this by also setting the MAX_CONNECTION_AGE_GRACE option to allow for a grace period to allow requests to finish.

I have a question about setting the MAX_CONNECTION_AGE_GRACE option.  Say that we have some request that is taking a very long time to finish, and holding up that specific connection.  No new requests are allowed to use the connection, correct? Is a new connect to that server opened up in the meantime? Or is all other communication with that specific server blocked until the requests in the grace period are complete?

**Comment by dfawley** [comment]
> I have a question about setting the MAX_CONNECTION_AGE_GRACE option. Say that we have some request that is taking a very long time to finish, and holding up that specific connection. No new requests are allowed to use the connection, correct?

That's correct.  When MAX_CONNECTION_AGE is hit, the server sends a GOAWAY to the client which prevents new streams from being created.  After MAX_CONNECTION_AGE_GRACE, it will forcibly close the connection.

> Is a new connect to that server opened up in the meantime?

Yes, a new connection to the server is (typically) created by the client when this happens.

**Comment by aarontami** [comment]
I see, i a bit confused at the behavior we are seeing.  Even with a very very large max age grace, we are seeing a burst of UNAVAILABLE errors on the client when connections expire.

**Comment by dfawley** [comment]
@aarontami what language? are you using pick-first so this is the only server you are connected to?  Or round robin? are your RPCs wait-for-ready or not-WFR?

**Comment by ejona86** [comment]
@aarontami, thank you for sharing your experience.

> I think we were able to solve this by also setting the MAX_CONNECTION_AGE_GRACE option to allow for a grace period to allow requests to finish.

This isn't likely. [The default grace time is infinite.](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md#proposal) So if anything this should cause more failures, not fewer.

> We started experiencing a large number of random UNAVAILABLE errors, which seemed to be due to the connection age and expiring and getting canceled while a request was in progress.

This is probably due to races on client-side between choosing a transport and sending the RPC. It seems Java's had the biggest problem historically here. But grpc-java v1.30 dramatically reduced the size of this race. From the release notes: "netty: Reduce race window size between GOAWAY and new streams."

You should share the full error you see though. I would expect UNAVAILABLE with GOAWAY and NO_ERROR. Depending on what your seeing, it could be something different.

**Comment by aarontami** [comment]
We are seeing the UNAVAILABLE errors with GOAWAY, so i think you are correct.  I will enable some more verbose logging and repro the problem to send you some better logs.

We are using a python client (using grpcio) and a few different servers, written in c++, go and python.  We see this error in all these cases when the MAX_CONNECTION_AGE is enabled.

**Comment by aarontami** [comment]
I enabled GRPC_VERBOSITY=DEBUG, I can enable GRPC_TRACE as well if you think it would be helpful:

```
I 2020-06-26T22:52:21.563277309Z   ipv4:10.21.252.12:7068: Got goaway [0] err={"created":"@1593211941.563258677","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.563538126Z   ipv4:10.21.243.16:7068: Got goaway [0] err={"created":"@1593211941.563522858","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.563753441Z   ipv4:10.21.210.154:7068: Got goaway [0] err={"created":"@1593211941.563737703","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.563880397Z   ipv4:10.21.224.29:7068: Got goaway [0] err={"created":"@1593211941.563863035","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.563950404Z   ipv4:10.21.173.40:7068: Got goaway [0] err={"created":"@1593211941.563933895","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.564022628Z   ipv4:10.21.214.41:7068: Got goaway [0] err={"created":"@1593211941.564009833","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.564140936Z         New connected subchannel at 0x7f455c289db0 for subchannel 0x7f455828dae0 
I 2020-06-26T22:52:21.564260297Z   ipv4:10.21.232.29:7068: Got goaway [0] err={"created":"@1593211941.564240693","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.564275481Z   ipv4:10.21.214.42:7068: Got goaway [0] err={"created":"@1593211941.564259259","description":"GOAWAY received","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":1162,"grpc_status":14,"http2_error":0,"raw_bytes":"max_age"} 
I 2020-06-26T22:52:21.564730915Z         New connected subchannel at 0x7f45740020d0 for subchannel 0x7f455801b880 
I 2020-06-26T22:52:21.564787173Z         New connected subchannel at 0x7f45483e3560 for subchannel 0x7f4558018ca0 
I 2020-06-26T22:52:21.565165722Z         New connected subchannel at 0x7f454825c3d0 for subchannel 0x7f4558016900 
I 2020-06-26T22:52:21.565221775Z         New connected subchannel at 0x7f455c0034f0 for subchannel 0x7f455828cc40 
I 2020-06-26T22:52:21.565294422Z         New connected subchannel at 0x7f4548374c00 for subchannel 0x7f455801a9e0 
E 2020-06-26T22:52:21.566028706Z [2020-06-26 22:52:21,565: ERROR/ForkPoolWorker-4] Retrying request in 0.5s due to StatusCode UNAVAILABLE, Error (14, 'unavailable'), Message Socket closed. Retries left 2
```

Here you can se the UNAVAILABLE error immediately following the GOAWAY signals.  We retry on this error on the client and usually it is fine, however sometimes we see a chain of these errors that continue to fail on retry.

**Comment by markdroth** [comment]
@aarontami Can you please file a separate issue about the UNAVAILABLE problem with MAX_CONNECTION_AGE on the server?  I suspect this is a known race condition that won't go away until we finish implementing transparent retries, but we can help you look at the logs to confirm that.

**Comment by hacst** [comment]
This issue has been open for quite a while now. I still think the issue is relevant today and is still creating problems in practice for grpc users. In my experience most users of grpc in kubernetes that do not go the service mesh/L7 LB route right away get tripped up by this. I just checked and the top hit that comes up for me on this topic on google still proposes DNS-LB using headless services and makes no mention of the required configuration (max connection age + grace) explained in this thread. It is fair to assume that there are users running with broken load balancing without even knowing it because of this.

For cluster internal communication in k8s and other server to server use-cases the number of clients and servers, except maybe in extremely rare cases, does not reach the point where the DNS scalability issues become a problem. Especially if mitigations like jitter were used.

While I'm not stuck on the idea of having to use DNS-LB to make grpc in kubernetes "just work" it still feels like the most obvious one to me.  Are there any other new options I am missing that make basic "proxyless" load balancing easy to get right in k8s across languages these days? Obviously what is right for this use-case might not be good for others but from my bubble it seems like a considerable amount of grpc usage even outside k8s these days goes along similar lines (s2s/k8s) and would benefit from something working out of the box there.

**Comment by joshua-barnett** [comment]
For anyone else getting stuck trying to implement load balancing the **_non-optimal_** (DNS) way I created this example project originally to help me debug scaling and round robin load balancing.
  
>[GRPC Node.js DNS Load Balancing Example](https://github.com/joshua-barnett/grpc-node-dns-lb)
>[![asciicast](https://asciinema.org/a/LpXY8cR0RbbgSeC5T3F6zwXTo.svg)](https://asciinema.org/a/LpXY8cR0RbbgSeC5T3F6zwXTo)
_**Note: the `grpc` module using the native add-on is currently pinned to v1.24.x of the gRPC C core.**_

I'm still not sure why the round robin load balancing fails to send requests sequentially to each replica, sometimes it will hit the same replica twice before sending a request to another replica.

**Comment by mfine** [comment]
The workaround suggested in the issue is to set ```MAX_CONNECTION_AGE_MS``` channel arguments on the server. Is there a workaround for long-running streaming services that don't have an applicable max connection age? Our client streams run "forever" and we'd like them to pick up added servers without disrupting them. Is there anything we can do to kick the client's channel to force it to re-resolve DNS names (we're happy to handle running this periodically outside of the library).

**Comment by markdroth** [comment]
There is no way to transparently move a stream from one server to another; you need to start a new stream for that.  In general, clients that use long-running streams need to be prepared to restart their stream any time it fails anyway, so you just need to arrange for the streams to be periodically terminated, and that will cause the streams to move to whatever new connections have been created by the clients.

One way to do this is to this is to set `MAX_CONNECTION_AGE_GRACE` on the server in addition to `MAX_CONNECTION_AGE_MS`, so that it will actually kill the existing connections after they timeout instead of just sending a GOAWAY and then waiting forever before terminating the connection.

Another option is to handle this at the application level by having your service implementation terminate streams after a certain amount of time.  That approach would need to be done individually for each service, so it may be more work if you have a large number of services, but it does allow the service implementation more control over shutting down the stream (e.g., it can send whatever status code or trailing metadata it wants).

**Comment by mfine** [comment]
@markdroth thanks for the response. We're more concerned about whether new streams we initiate from a client pick up the added servers and would like to avoid disrupting the existing streams from the client. If our client has streams 1, 2, and 3 connected to server A, when server B comes up we'd like the new stream 4 to consider both servers A and B without having to disrupt streams 1, 2, and 3 connected to server A.

**Comment by markdroth** [comment]
That will work properly with the approach I've described.  Once the client sees a GOAWAY on an existing connection, it will stop using that connection for new streams, even if the connection remains open for a long time for the existing streams that had already been started on that connection before the GOAWAY was received.  The GOAWAY will trigger the client to stop using that connection and to do a DNS re-resolution, at which point it will pick up the new addresses, and all of the new addresses will be considered for subsequently started streams.

**Comment by mfine** [comment]
@markdroth if we were to set `MAX_CONNECTION_AGE_MS` but then leave `MAX_CONNECTION_AGE_GRACE_MS` to the default `infinite`, the client would stop using the connection for new streams but never actually get rid of any existing streams on the connection? Could that not lead to accumulating connections? Like in my scenario above, streams 1, 2, and 3 are connected to server A on connection i - then new stream 4 considers both server A and B and ends up connecting to server A on connection ii.

**Comment by markdroth** [comment]
Yes, that could lead to accumulating connections.  If that's a concern, it can be ameliorated by either setting `MAX_CONNECTION_AGE_GRACE_MS` or by having the application terminate streams after some fixed period of time, as I mentioned above.

**Comment by mqsoh** [comment]
The Go client has [this resolver](https://pkg.go.dev/google.golang.org/grpc/resolver) thing. You can import the package and Register a resolver builder or whatever. The client automatically picks it up, so I guess it's a global.

My coworker hacked something together to watch endpoints in the Kubernetes API and we have perfect load balancing when pods go in an out of service. I got here trying to figure out how to do it with the Node.js client. I guess it's not possible.

Although, this repo isn't specific to the Node.js client and at least one other language client already does this. I'm confused.

**Comment by ejona86** [comment]
@mqsoh, please file a separate issue. This issue is about when a name resolver refreshes itself (when polling is necessary). You want a custom name resolver API in Node.js. Your request is more similar to https://github.com/grpc/grpc/issues/11879 and https://github.com/grpc/grpc/issues/11685

**Comment by mqsoh** [comment]
I'm not reporting an issue. I was trying to help by pointing out that a different client already has a solution to OP's issue.

The issues you linked me to seem to be other clients that, along with the Node.js client, also lack features developed for the Go client.

**Comment by ejona86** [comment]
@mqsoh, oh, okay. Yes, many custom resolvers wouldn't have the problem discussed here as many naming systems have watch APIs. When available in your language and it makes sense for your deployment, such a custom/pluggable name resolver would be the preferred solution. And there is (slow) progress on allowing custom name resolvers in more languages. I don't think there is much argument there.

I think this discussion is more focusing on where you _don't_ have a watch API (aka DNS), or even though a watch API might be available you aren't in direct control of the clients and so practically you need to use DNS.

**Comment by Sovietaced** [comment]
> The Go client has [this resolver](https://pkg.go.dev/google.golang.org/grpc/resolver) thing. You can import the package and Register a resolver builder or whatever. The client automatically picks it up, so I guess it's a global.
> 
> My coworker hacked something together to watch endpoints in the Kubernetes API and we have perfect load balancing when pods go in an out of service. I got here trying to figure out how to do it with the Node.js client. I guess it's not possible.

From what I understand this still isn't a silver bullet because the pluggable resolver is only called when a channel (or a sub channel now?) fails. Therefore if your target service is only scaling up, your client will not distribute load to new pods.

**Comment by mqsoh** [comment]
> > The Go client has [this resolver](https://pkg.go.dev/google.golang.org/grpc/resolver) thing. You can import the package and Register a resolver builder or whatever. The client automatically picks it up, so I guess it's a global.
> > My coworker hacked something together to watch endpoints in the Kubernetes API and we have perfect load balancing when pods go in an out of service. I got here trying to figure out how to do it with the Node.js client. I guess it's not possible.
> 
> From what I understand this still isn't a silver bullet because the pluggable resolver is only called when a channel (or a sub channel now?) fails. Therefore if your target service is only scaling up, your client will not distribute load to new pods.

A project maintainer already marked my comment as irrelevant two and a half years ago.

**Comment by struckoff** [comment]
I know someone uses KeepAlive policy to fore rediscovery, since I dont wanna do it on client side, is it safe just trigger d.watcher (or dnsResolver.ResolveNow)with a timer, every 30s for example?

**Comment by nancyxu825** [comment]
We ran into this issue in a Cloud Kubernetes environment during pod rolling restarts. When a backend pod rolls and gets a new IP, the old IP blackholes incoming traffic (no immediate TCP RST).

Sequence of events: 
1. READY: Client is connected to Server at IP_1.
2. GOAWAY: Server gracefully shuts down, sends GOAWAY, and disconnects. Channel enters IDLE.
3. Disconnect DNS check: gRPC triggers one resolution upon disconnect; because the server is still shutting down, DNS still returns IP_1. Resolution succeeds, so gRPC stops querying.
4. IP Change during IDLE: Server terminates and restarts on IP_2. DNS updates to IP_2. Client channel can remain in IDLE for an arbitrary amount of time, retaining the stale IP
5. Stall: When a new RPC arrives, gRPC attempts to connect to cached IP_1 (now a blackhole), stalling for the full GRPC_ARG_MIN_RECONNECT_BACKOFF_MS timeout before failing over, re-resolving DNS, and connecting to IP_2

Tuning GRPC_ARG_MIN_RECONNECT_BACKOFF_MS to a lower value is not necessarily a viable fix because it can potentially cause spurious connection failures on legitimate slow connections.

Current Workaround: We monitor non-ready DNS channels and re-create the channel upon detection of the channel having a stale IP.

It would be a huge improvement if we could trigger a fresh re-resolution upon exiting IDLE.
